### PR TITLE
Hide `ResumeOverlay` when `OsuModAutopilot` is enabled

### LINKED ISSUE #22001: Mod "Autopilot" shouldn't display click-to-resume overlay
### Type

Game behaviour

### Bug description

Stable doesn't display this overlay when the mod is on. And generally, the cursor's position doesn't matter when resuming gameplay in such case, so there's no reason to have it displayed.

### Screenshots or videos

_No response_

### Version

2022.1228.0

### Logs

-

### PR DESCRIPTION
**Pull request by Pasi4K5**
Fixes #22001

### CODE REVIEW

File: osu.Game/Rulesets/Mods/Mod.cs
Comment (bdach) [inline osu.Game/Rulesets/Mods/Mod.cs:111]: This cannot exist here. `Mod` is the wrong level of abstraction for this. All mods implement their effects via `IApplicableMod` interfaces, rather than putting stuff into `Mod` directly. All overridable properties on `Mod` are more like metadata about the mod rather and they do not generally control its operation.

File: osu.Game/Rulesets/UI/DrawableRuleset.cs
Comment (bdach) [inline osu.Game/Rulesets/UI/DrawableRuleset.cs:235]: I would explore the possibility of doing something so that `DrawableRuleset` can preview the active mods at the point of creating the resume overlay, and *not* create it if it determines that the active mods are incompatible with resume. It would remove the requirement of jamming even more ugly logic in here.

Note: I'm not saying to *do this* explicitly, I'm saying to *explore* it. As in: consider it, maybe try how it looks code-wise, and then push *if* the direction ends up making sense.

**Review by bdach — CHANGES_REQUESTED** [review]
No tests for this?

File: osu.Game/Rulesets/Mods/Mod.cs
Comment (Pasi4K5) [reply osu.Game/Rulesets/Mods/Mod.cs]: I removed the prop and set `drawableRuleset.ResumeOverlay` to `null` in `ApplyToDrawableRuleset()`.

File: osu.Game/Rulesets/UI/DrawableRuleset.cs
Comment (Pasi4K5) [reply osu.Game/Rulesets/UI/DrawableRuleset.cs]: Not really possible with the approach I took because the `ResumeOverlay` is set in `ApplyToDrawableRuleset` after it's already been created. But the excess ugly logic from before is gone now anyway.

File: osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs:57]: This just cannot. I'm not sure even where to start with how wrong this is. You're setting a property to `null` without expiring, disposing, or anything. I don't even know how this would possibly work.

**Comment by peppy** [comment]
This needs test coverage.

File: osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
Comment (Pasi4K5) [reply osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs:57]: > I don't even know how this would possibly work.

Well, it works because `ResumeOverlay` being `null` is handled accordingly in `DrawableRuleset`.

Anyway, would expiring the `ResumeOverlay` instead and checking `ResumeOverlay.IsAlive` here in `RequestResume()` be an acceptable solution?
https://github.com/ppy/osu/blob/8b47af6503b5a93e98940d31b96d13b1858361f5/osu.Game/Rulesets/UI/DrawableRuleset.cs#L231-L241

Comment (peppy) [reply osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs:57]: I don't think that's a good direction.

I'd see this working something like:

```diff
diff --git a/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs b/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
index 6772cfe0be..ad15177898 100644
--- a/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
+++ b/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
@@ -23,7 +23,9 @@ public class OsuModAutopilot : Mod, IApplicableFailOverride, IUpdatableByPlayfie
         public override ModType Type => ModType.Automation;
         public override LocalisableString Description => @"Automatic cursor movement - just follow the rhythm.";
         public override double ScoreMultiplier => 0.1;
-        public override Type[] IncompatibleMods => new[] { typeof(OsuModSpunOut), typeof(ModRelax), typeof(ModFailCondition), typeof(ModNoFail), typeof(ModAutoplay), typeof(OsuModMagnetised), typeof(OsuModRepel) };
+
+        public override Type[] IncompatibleMods => new[]
+            { typeof(OsuModSpunOut), typeof(ModRelax), typeof(ModFailCondition), typeof(ModNoFail), typeof(ModAutoplay), typeof(OsuModMagnetised), typeof(OsuModRepel) };
 
         public bool PerformFail() => false;
 
@@ -60,6 +62,8 @@ public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset
 
             // Generate the replay frames the cursor should follow
             replayFrames = new OsuAutoGenerator(drawableRuleset.Beatmap, drawableRuleset.Mods).Generate().Frames.Cast<OsuReplayFrame>().ToList();
+
+            drawableRuleset.UseResumeOverlay = false;
         }
     }
 }
diff --git a/osu.Game/Rulesets/UI/DrawableRuleset.cs b/osu.Game/Rulesets/UI/DrawableRuleset.cs
index 71b452c309..d5a4ce069d 100644
--- a/osu.Game/Rulesets/UI/DrawableRuleset.cs
+++ b/osu.Game/Rulesets/UI/DrawableRuleset.cs
@@ -196,6 +196,8 @@ private void load(CancellationToken? cancellationToken)
 
             if ((ResumeOverlay = CreateResumeOverlay()) != null)
             {
+                UseResumeOverlay = true;
+
                 AddInternal(CreateInputManager()
                     .WithChild(CreatePlayfieldAdjustmentContainer()
                         .WithChild(ResumeOverlay)));
@@ -230,7 +232,7 @@ private void loadObjects(CancellationToken cancellationToken)
 
         public override void RequestResume(Action continueResume)
         {
-            if (ResumeOverlay != null && (Cursor == null || (Cursor.LastFrameState == Visibility.Visible && Contains(Cursor.ActiveCursor.ScreenSpaceDrawQuad.Centre))))
+            if (ResumeOverlay != null && UseResumeOverlay && (Cursor == null || (Cursor.LastFrameState == Visibility.Visible && Contains(Cursor.ActiveCursor.ScreenSpaceDrawQuad.Centre))))
             {
                 ResumeOverlay.GameplayCursor = Cursor;
                 ResumeOverlay.ResumeAction = continueResume;
@@ -507,6 +509,12 @@ internal DrawableRuleset(Ruleset ruleset)
         /// </summary>
         public ResumeOverlay ResumeOverlay { get; protected set; }
 
+        /// <summary>
+        /// Whether the <see cref="ResumeOverlay"/> should be used to return the user's cursor position to its previous location after a pause.
+        /// </summary>
+        /// <remarks>Defaults to <c>true</c> if a ruleset returns a non-null overlay via <see cref="CreateResumeOverlay"/>.</remarks>
+        public bool UseResumeOverlay { get; set; }
+
         /// <summary>
         /// Returns first available <see cref="HitWindows"/> provided by a <see cref="HitObject"/>.
         /// </summary>

```

Let me know if you want to add a test and update to use this direct, else I can PR an alternative for you.

Comment (Pasi4K5) [reply osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs:57]: I think it would be better if we did it directly in `Player` though. This way it's handled in the same way as an instant resume during a break.

```diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -82,6 +82,11 @@ public abstract partial class Player : ScreenWithBeatmapBackground, ISamplePlaybackDisabler, ILocalUserPlayInfo

         /// </summary>
         protected virtual bool PauseOnFocusLost => true;

+        /// <summary>
+        /// Always resumes instantly without showing a <see cref="ResumeOverlay"/>.
+        /// </summary>
+        public bool AlwaysInstantResume { get; set; }
+
         public Action<bool> RestartRequested;

         private bool isRestarting;
@@ -978,8 +983,8 @@ public void Resume()
             IsResuming = true;
             PauseOverlay.Hide();

-            // breaks and time-based conditions may allow instant resume.
-            if (breakTracker.IsBreakTime.Value)
+            // breaks and time-based conditions as well as some mods may allow instant resume.
+            if (breakTracker.IsBreakTime.Value || AlwaysInstantResume)
                 completeResume();
             else
                 DrawableRuleset.RequestResume(completeResume);
--- a/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
+++ b/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
@@ -12,10 +12,11 @@
 using osu.Game.Rulesets.Osu.Objects;
 using osu.Game.Rulesets.Osu.Replays;
 using osu.Game.Rulesets.UI;
+using osu.Game.Screens.Play;

 namespace osu.Game.Rulesets.Osu.Mods
 {
-    public class OsuModAutopilot : Mod, IApplicableFailOverride, IUpdatableByPlayfield, IApplicableToDrawableRuleset<OsuHitObject>
+    public class OsuModAutopilot : Mod, IApplicableFailOverride, IUpdatableByPlayfield, IApplicableToDrawableRuleset<OsuHitObject>, IApplicableToPlayer
     {
         public override string Name => "Autopilot";
         public override string Acronym => "AP";
@@ -63,5 +64,10 @@ public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset
             // Generate the replay frames the cursor should follow
             replayFrames = new OsuAutoGenerator(drawableRuleset.Beatmap, drawableRuleset.Mods).Generate().Frames.Cast<OsuReplayFrame>().ToList();
         }
+
+        public void ApplyToPlayer(Player player)
+        {
+            player.AlwaysInstantResume = true;
+        }
     }
 }
```
Plus, it doesn't make this condition even longer than it already is.
https://github.com/ppy/osu/blob/8b47af6503b5a93e98940d31b96d13b1858361f5/osu.Game/Rulesets/UI/DrawableRuleset.cs#L233

**Comment by peppy** [comment]
I've moved the tests over to a `ModTestScene` to conform with other mod tests.

I've also restructured the `UseResumeOverlay` flag so in the potential edge case of it being set to `false` before the `DrawableRuleset` it loaded, it will still handle correctly.

Would appreciate a second opinion on this before merge.

**Review by peppy — DISMISSED** [review]

File: osu.Game/Rulesets/UI/DrawableRuleset.cs
Comment (bdach) [inline osu.Game/Rulesets/UI/DrawableRuleset.cs:200]: @peppy I'm not getting this. You've mentioned an edge case when the flag is set before BDL, but... why does it matter? What's the fail case?

If the flag is set to false before BDL, then the overlay will be created and added to the hierarchy - but it *should* never be shown, because the `RequestResume()` check tests both the presence of the resume overlay and the value of the flag - the former may be non-null, but the flag is still false?

Is all this just for optimisation? What am I missing here? I wouldn't bother asking but I find it really weird that the drawable rulesest is setting a "config" bool flag on self - and even then, what does that achieve if you could, for instance, set it back to `true` at any point post-load? And if you did, then this implementation is arguably *less* correct, because as the flag is only tested once in BDL, any value changes to it post-BDL effectively don't matter, which make it a property whose behaviour is dependent on timing?

I've also checked out one commit back (https://github.com/ppy/osu/pull/22338/commits/9e04a36d86c9a19e37cf470aa4c4505fa9ad4ef9) and the test still passes, which adds to my confusion.

Comment (peppy) [reply osu.Game/Rulesets/UI/DrawableRuleset.cs:200]: There's no test coverage of this, but the previous logic would set `UseResumeOverlay` to `true` as long as `CreateResumeOverlay` could be created.

My point is that if there's a future *non-mod* usage that sets the property before `LoadComponent`, it would fail and be overwritten by the code in `load()`.

ie.

```csharp
var ruleset = new DrawableOsuRuleset()
{
    UseResumeOverlay = false
};

LoadComponent(ruleset);

Debug.Assert(!ruleset.UseResumeOverlay); // would have failed
```

Comment (bdach) [reply osu.Game/Rulesets/UI/DrawableRuleset.cs:200]: Thanks, I missed that part. For what it's worth I also find the previous variant of that code weird, and don't really get why the creation of the overlay *must* somehow intermingle with the value of the flag in either direction, so I guess I'll need to go stare at it some more and try to find out...

Comment (peppy) [reply osu.Game/Rulesets/UI/DrawableRuleset.cs:200]: I believe that part is just an optimisation, or more a correctness factor (a variable called `UseResumeOverlay` should probably not be `true` if there's no resume overlay). I'm open to removing that cross-talk if you prefer.

Comment (bdach) [reply osu.Game/Rulesets/UI/DrawableRuleset.cs:200]: > I'm open to removing that cross-talk if you prefer.

I think I'd prefer that. I don't find `UseResumeOverlay` having no effect if no overlay was created _that_ weird - definitely not as weird as "spooky action at a distance" timing-dependent flag value changes.

Comment (peppy) [reply osu.Game/Rulesets/UI/DrawableRuleset.cs]: I've adjusted, and also removed the shortcutting when `UseResumeOverlay` is false. Let me know if you agree with that.

**Review by bdach — APPROVED** [review]

**Pull request merged**
